covenant-signer/itest/e2e_test.go

```diff
@@ -0,0 +1,352 @@
+	"github.com/babylonlabs-io/babylon/testutil/datagen"
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/stretchr/testify/require"
+
+	"github.com/babylonlabs-io/covenant-emulator/covenant-signer/config"
+	"github.com/babylonlabs-io/covenant-emulator/covenant-signer/itest/containers"
+	"github.com/babylonlabs-io/covenant-emulator/covenant-signer/keystore/cosmos"
+	"github.com/babylonlabs-io/covenant-emulator/covenant-signer/observability/metrics"
+	"github.com/babylonlabs-io/covenant-emulator/covenant-signer/signerapp"
+	"github.com/babylonlabs-io/covenant-emulator/covenant-signer/signerservice"
+	"github.com/babylonlabs-io/covenant-emulator/covenant-signer/signerservice/types"
+)
+
+var (
+	netParams              = &chaincfg.RegressionNetParams
+	eventuallyPollInterval = 100 * time.Millisecond
+	eventuallyTimeout      = 10 * time.Second
+)
+
+type TestManager struct {
+	t               *testing.T
+	bitcoindHandler *BitcoindTestHandler
+	walletPass      string
+	covenantPrivKey *btcec.PrivateKey
+	signerConfig    *config.Config
+	app             *signerapp.SignerApp
+	server          *signerservice.SigningServer
+}
+
+type stakingData struct {
+	stakingAmount  btcutil.Amount
+	stakingTime    uint16
+	stakingFeeRate btcutil.Amount
+}
+
+func defaultStakingData() *stakingData {
+	return &stakingData{
+		stakingAmount:  btcutil.Amount(100000),
+		stakingTime:    10000,
+		stakingFeeRate: btcutil.Amount(5000), // feeRatePerKb
+	}
+}
+
+func StartManager(
+	t *testing.T,
+	numMatureOutputsInWallet uint32) *TestManager {
+	m, err := containers.NewManager()
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		_ = m.ClearResources()
+	})
+
+	h := NewBitcoindHandler(t, m)
+	h.Start()
+
+	// Give some time to launch and bitcoind
+	time.Sleep(2 * time.Second)
+
+	passphrase := "pass"
+	_ = h.CreateWallet("test-wallet", passphrase)
+	// only outputs which are 100 deep are mature
+	_ = h.GenerateBlocks(int(numMatureOutputsInWallet) + 100)
+
+	appConfig := config.DefaultConfig()
+
+	appConfig.KeyStore.KeyStoreType = "cosmos"
+	appConfig.KeyStore.CosmosKeyStore.ChainID = "test-chain"
+	appConfig.KeyStore.CosmosKeyStore.Passphrase = passphrase
+	appConfig.KeyStore.CosmosKeyStore.KeyName = "test-key"
+	appConfig.KeyStore.CosmosKeyStore.KeyDirectory = ""
+	appConfig.KeyStore.CosmosKeyStore.KeyringBackend = "memory"
+
+	retriever, err := cosmos.NewCosmosKeyringRetriever(appConfig.KeyStore.CosmosKeyStore)
+	require.NoError(t, err)
+
+	keyInfo, err := retriever.Kr.CreateChainKey(
+		appConfig.KeyStore.CosmosKeyStore.Passphrase,
+		"",
+	)
+	require.NoError(t, err)
+
+	app := signerapp.NewSignerApp(
+		retriever,
+	)
+
+	met := metrics.NewCovenantSignerMetrics()
+	parsedConfig, err := appConfig.Parse()
+	require.NoError(t, err)
+
+	server, err := signerservice.New(
+		context.Background(),
```

Comment:
user proper context

Reply:
agreed on context related stuff, for now I have created issue to address all context related improvements in separate pr - https://github.com/babylonlabs-io/covenant-emulator/issues/34